sdk/dotnet/Pulumi.Automation/Events/EventLogWatcher.cs

```diff
@@ -30,94 +29,74 @@ string logFile
-                }
-            }
-
-            // Try reading events one final time
             try
             {
-                ReadEvents();
+                await this._pollingTask;
             }
-            catch
+            catch (OperationCanceledException error) when (error.CancellationToken == this._cancellationToken)
             {
-                // ignored
+                // _pollingTask.State == Cancelled
             }
         }
 
-        private void ReadEvents()
+        private async Task PollForEvents(CancellationToken externalCancellationToken)
         {
-            if (!File.Exists(LogFile))
-            {
-                return;
-            }
+            using var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(
+                this._internalCancellationTokenSource.Token,
+                externalCancellationToken);
+            this._cancellationToken = linkedSource.Token;
 
-            using var fs = new FileStream(LogFile, FileMode.Open, FileAccess.Read);
-            var newLength = fs.Length;
+            await ReadEventsOnce();
 
-            if (newLength == _previousLength)
+            // At least one non-interruptible delay to ensure the thread has a chance
+            // to read just-written data.
+            await Task.Delay(_pollingIntervalMilliseconds); 
+
+            while (true)
             {
-                return;
+                await ReadEventsOnce();
+                await Task.Delay(_pollingIntervalMilliseconds, linkedSource.Token);
             }
+        }
 
-            fs.Seek(_previousLength, SeekOrigin.Begin);
-
+        private async Task ReadEventsOnce()
+        {
+            using var fs = new FileStream(LogFile, FileMode.Open, FileAccess.Read)
+            {
+                Position = this._position
+            };
             using var reader = new StreamReader(fs);
-
-            var lines = reader
-                .ReadToEnd()
-                .Split(Environment.NewLine.ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
-
-            foreach (var line in lines)
+            string? line;
+            while (reader.Peek() >= 0)
             {
-                var @event = _localSerializer.DeserializeJson<EngineEvent>(line);
-
-                try
+                line = await reader.ReadLineAsync();
+                this._position = fs.Position;
+                if (!String.IsNullOrWhiteSpace(line))
                 {
+                    line = line.Trim();
+                    var @event = _localSerializer.DeserializeJson<EngineEvent>(line);
                     _onEvent.Invoke(@event);
                 }
-                catch
-                {
-                    // Don't let the provided event handler cause reading of events to fail
-                }
             }
+        }
 
-            _previousLength = newLength;
+        public void Dispose()
+        {
+            ((IDisposable)_internalCancellationTokenSource).Dispose();
```

Comment:
This cast should not be necessary

```suggestion
            _internalCancellationTokenSource.Dispose();
```